src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala

```diff
@@ -362,12 +602,481 @@ object FunctionRegistry {
+      def typ: Type = hrt.typ
+    }
+
+    register("<", ord.lt _)
+    register("<=", ord.lteq _)
+    register(">", ord.gt _)
+    register(">=", ord.gteq _)
+
+    registerMethod("min", ord.min _)
+    registerMethod("max", ord.max _)
+
+    registerMethod("sort", (a: IndexedSeq[Any]) => a.sorted(extendOrderingToNull(ord)))(arrayHr(hrboxedt), arrayHr(hrboxedt))
+    registerMethod("sort", (a: IndexedSeq[Any], ascending: Boolean) =>
+      a.sorted(extendOrderingToNull(
+        if (ascending)
+          ord
+        else
+          ord.reverse))
+    )(arrayHr(hrboxedt), boolHr, arrayHr(hrboxedt))
+
+    registerLambdaMethod("sortBy", (a: IndexedSeq[Any], f: (Any) => Any) =>
+      a.sortBy(f)(extendOrderingToNull(ord))
+    )(arrayHr(TTHr), unaryHr(TTHr, hrboxedt), arrayHr(TTHr))
+
+    registerLambdaMethod("sortBy", (a: IndexedSeq[Any], f: (Any) => Any, ascending: Boolean) =>
+      a.sortBy(f)(extendOrderingToNull(
+        if (ascending)
+          ord
+        else
+          ord.reverse))
+    )(arrayHr(TTHr), unaryHr(TTHr, hrboxedt), boolHr, arrayHr(TTHr))
+  }
+
+  registerOrderedType[Int]()
+  registerOrderedType[Long]()
+  registerOrderedType[Float]()
+  registerOrderedType[Double]()
+  registerOrderedType[String]()
+
+  register("%", (x: Int, y: Int) => x % y)
+  register("%", (x: Long, y: Long) => x % y)
+  register("+", (x: String, y: Any) => x + y)(stringHr, TTHr, stringHr)
+
+  register("~", (s: String, t: String) => s.r.findFirstIn(t).isDefined)
+
+  registerSpecial("isMissing", (g: () => Any) => g() == null)(TTHr, boolHr)
+  registerSpecial("isDefined", (g: () => Any) => g() != null)(TTHr, boolHr)
+
+  registerSpecial("json", (f: () => Any) => JsonMethods.compact(TT.t.toJSON(f())))(TTHr, stringHr)
+  registerSpecial("str", (f: () => Any) => TT.t.str(f()))(TTHr, stringHr)
+
+  registerSpecial("||", { (f1: () => Any, f2: () => Any) =>
+    val x1 = f1()
+    if (x1 != null) {
+      if (x1.asInstanceOf[Boolean])
+        true
+      else
+        f2()
+    } else {
+      val x2 = f2()
+      if (x2 != null
+        && x2.asInstanceOf[Boolean])
+        true
+      else
+        null
+    }
+  })(boolHr, boolHr, boxedboolHr)
+
+  registerSpecial("&&", { (f1: () => Any, f2: () => Any) =>
+    val x = f1()
+    if (x != null) {
+      if (x.asInstanceOf[Boolean])
+        f2()
+      else
+        false
+    } else {
+      val x2 = f2()
+      if (x2 != null
+        && !x2.asInstanceOf[Boolean])
+        false
+      else
+        null
+    }
+  })(boolHr, boolHr, boxedboolHr)
+
+  registerMethodSpecial("orElse", { (f1: () => Any, f2: () => Any) =>
+    val v = f1()
+    if (v == null)
+      f2()
+    else
+      v
+  })(TTHr, TTHr, TTHr)
+
+  registerMethod("[]", (a: IndexedSeq[Any], i: Int) => a(i))(arrayHr(TTHr), intHr, TTHr)
+  registerMethod("[]", (a: Map[String, Any], i: String) => a(i))(dictHr(TTHr), stringHr, TTHr)
+  registerMethod("[]", (a: String, i: Int) => a(i).toString)(stringHr, intHr, stringHr)
```

Comment:
Fixed.

Good question.  Let's leave it for another day.  (Although probably not.  It doesn't get full treatment and I don't think anybody uses it.)